pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,455 @@
+	p := &podpools{
+		Logger:       logger.NewLogger(PolicyName),
+		options:      policyOptions,
+		cpuAllocator: cpuallocator.NewCPUAllocator(policyOptions.System),
+	}
+
+	p.Info("creating policy...")
+	p.allowed = policyOptions.Available[policyapi.DomainCPU].(cpuset.CPUSet)
+	p.reserved = cpuset.NewCPUSet()
+	if reserved, ok := p.options.Reserved[policyapi.DomainCPU]; ok {
+		switch v := reserved.(type) {
+		case cpuset.CPUSet:
+			p.reserved = p.allowed.Intersection(v)
+		case resapi.Quantity:
+			reserveCnt := (int(v.MilliValue()) + 999) / 1000
+			cpus, err := p.cpuAllocator.AllocateCpus(&p.allowed, reserveCnt, false)
+			if err != nil {
+				p.Fatal("failed to allocate reserved CPUs: %s", err)
+			}
+			p.reserved = cpus
+			p.allowed = p.allowed.Union(cpus)
+		}
+	}
+	if p.reserved.IsEmpty() {
+		p.Fatal("podpools cannot run without reserved CPUs that are also AvailableResources")
+	}
+	p.reservedMems = p.closestMems(p.reserved)
+	p.Debug("creating configuration: --- %+v ---", p.options.Available)
+	if err := p.setConfig(podpoolsOptions); err != nil {
+		p.Fatal("failed to create podpools policy: %v", err)
+	}
+
+	pkgcfg.GetModule(PolicyPath).AddNotify(p.configNotify)
+
+	return p
+}
+
+// Name returns the name of this policy.
+func (p *podpools) Name() string {
+	return PolicyName
+}
+
+// Description returns the description for this policy.
+func (p *podpools) Description() string {
+	return PolicyDescription
+}
+
+// Start prepares this policy for accepting allocation/release requests.
+func (p *podpools) Start(add []cache.Container, del []cache.Container) error {
+	p.Debug("got started...")
+	return nil
+}
+
+// Sync synchronizes the active policy state.
+func (p *podpools) Sync(add []cache.Container, del []cache.Container) error {
+	p.Debug("synchronizing state...")
+	for _, c := range del {
+		p.ReleaseResources(c)
+	}
+	for _, c := range add {
+		p.AllocateResources(c)
+	}
+	return nil
+}
+
+// AllocateResources is a resource allocation request for this policy.
+func (p *podpools) AllocateResources(c cache.Container) error {
+	p.Debug("allocating container %s...", c.PrettyName())
+	// Assign container to correct pool.
+	if pool := p.allocatePool(c); pool != nil {
+		p.Debug("assigning container %s to pool %s[%d]", c.PrettyName(), pool.Type.Name, pool.Instance)
+		podID := c.GetPodID()
+		pool.PodIDs[podID] = append(pool.PodIDs[podID], c.GetCacheID())
+		p.pinCpuMem(c, pool.CPUs, pool.Mems)
+		p.Debug("pool: %v", pool)
+	} else {
+		// Cannot assign container to any of the pooled CPUs.
+		// Assign to reserved or shared CPUs.
+		if p.reserved.Size() > 0 && (c.GetNamespace() == "kube-system" || p.shared.Size() == 0) {
+			p.Debug("pinning container %s to reserved CPUs", c.PrettyName())
+			p.pinCpuMem(c, p.reserved, p.reservedMems)
+		} else if p.shared.Size() > 0 {
+			p.Debug("pinning container %s to shared CPUs", c.PrettyName())
+			p.pinCpuMem(c, p.shared, p.sharedMems)
```

Comment:
While the kubelet has already set this for the container, the other policies follow the principle of setting all container parameters that they want to have set.

Reply:
Thanks @klihub, this definitely makes sense when containers is pinned to `reserved` or `shared` cpusets. Looking into it...

Reply:
Fixed. Added a "QoS test" that checks CPU shares value for all service classes. (The test has pretty general name because I believe there will be other checks for combinations of {BestEffort, Burstable, Guaranteed} x {pool, reserved, shared}, too.)